G5 X Y Z Babystepping

gralco opened the pull request:
G5 X Y Z — The argument is the number of steps to be performed in the positive or negative direction of the given axes (in the order Z, X, then Y). It attempts to prevent endstop crashes to the best of its ability. **\* Added the option to babystep the stored home_offset[Z_AXIS] if BABYSTEP_OFFSET is defined (this can also be performed on an LCD controller, and saved to the EEPROM as with standard home-offsets) ***

gralco committed:
Initial G5 X Y Z Babystepping

gralco committed:
Initial G5 X Y Z Babystepping

gralco committed:
If multiple axes are given then perform each of them in the following order: Z, X, then Y.

gralco committed:
If multiple axes are given then perform each of them in the following order: Z, X, then Y.

gralco committed:
Undefine BABYSTEPPING in Configuration_adv.h

gralco committed:
Undefine BABYSTEPPING in Configuration_adv.h

gralco committed:
Keep track of distance from endstop if babystepping from an LCD controller. Also enable setting babystepping home_offset[Z_axis] from the LCD.

gralco committed:
Keep track of distance from endstop if babystepping from an LCD controller. Also enable setting babystepping home_offset[Z_axis] from the LCD.

gralco committed:
Undefine BABYSTEP_OFFSET by default and included a description.

gralco committed:
Undefine BABYSTEP_OFFSET by default and included a description.

gralco committed:
More description.
Just ignoring incorrect code values.

gralco committed:
More description.
Just ignoring incorrect code values.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -1058,7 +1059,11 @@ static void axis_is_at_home(AxisEnum axis) {
   {
-    current_position[axis] = base_home_pos(axis) + home_offset[axis];
+    current_position[axis] = base_home_pos(axis)
+    #ifndef BABYSTEPPING
```
Should this be `BABYSTEP_OFFSET` here?

gralco replied:
Yup, you're right, just changed it, nice catch.

gralco committed:
Should be BABYSTEP_OFFSET.

gralco committed:
Should be BABYSTEP_OFFSET.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -1713,6 +1718,26 @@ static void homeaxis(AxisEnum axis) {
+    }
+    #endif //BABYSTEP_XY
+    #endif //BABYSTEPPING
+
```
Indentation should be applied to all blocks, including preprocessor blocks. There should be a space after `if`. And, braces can be placed at the end of the line…

```
#ifdef BABYSTEPPING
  if (axis == Z_AXIS) {
    baby_max_endstop[axis] = Z_BABY_DEFAULT_MAX_POS;
    baby_min_endstop[axis] = Z_BABY_DEFAULT_MIN_POS;
  }
  #ifdef BABYSTEP_XY
    else if (axis == X_AXIS) {
      baby_max_endstop[axis] = X_BABY_DEFAULT_MAX_POS;
      baby_min_endstop[axis] = X_BABY_DEFAULT_MIN_POS;
    }
    else if (axis == Y_AXIS) {
      baby_max_endstop[axis] = Y_BABY_DEFAULT_MAX_POS;
      baby_min_endstop[axis] = Y_BABY_DEFAULT_MIN_POS;
    }
  #endif //BABYSTEP_XY
#endif //BABYSTEPPING
```

gralco replied:
I added the indentation, I personally prefer the open curly brackets to have their own line, I think it makes the blocks/nests clearer.

Wackerbarth replied:
> On May 28, 2015, at 10:34 AM, gralco notifications@github.com wrote:
> I added the indentation, I personally prefer the open curly brackets to have their own line, I think it makes the blocks/nests clearer.

Rather than relying on “personal preference”, the Marlin project should adopt a standard.

Without addressing “which standard”, it should be something that `astyle` understands.

Then we should reformat ALL of the code according to that style and, in the future, accept merges only after they have been “cleaned up” to meet the standard. (The Travis tests can include a check to verify that running astyle against the code produces no changes)

gralco replied:
I can agree with that, but if you want PRs that don't include my personal preference and for me to follow the aesthetic standard then maybe add me to the Marlin dev team, it'll give me a sense of duty ;)

What's the procedure to gain write access anyway?

AnHardt replied:
We already have a coding-standard and try to get the code conforming to it step by step.
http://www.marlinfirmware.org/index.php/Coding_Standards

@gralco 
Show what you can - for a while. Than ask.

Wackerbarth replied:
> We already have a coding-standard and try to get the code conforming to it step by step.
> http://www.marlinfirmware.org/index.php/Coding_Standards http://www.marlinfirmware.org/index.php/Coding_StandardsI’ve glad to see that you have a standard. My next question is “Is that a configuration file for `astyle` that matches this standard (particularly WRT indents and block delimiters)? If not, I suggest that you choose a standard that can be so described. That way, as a part of your Travis checking, you can also see if the style is followed and raise some alarm if it is not.

I would also like to see the standard specify that there be no whitespace at the end of lines.

gralco replied:
@AnHardt
Ah I see now, thanks for sharing!

thinkyhead replied:
> no whitespace at the end of lines

That almost goes without saying, but I will add it to the guidelines.

gralco committed:
Format blocks for preprocessor conditionals.

gralco committed:
Format blocks for preprocessor conditionals.

gralco committed:
Only babystep offset for Z

gralco committed:
Only babystep offset for Z

gralco committed:
Should be home_offset

gralco committed:
Should be home_offset

gralco committed:
These don't need to be volatile.

gralco committed:
These don't need to be volatile.

gralco committed:
Also shouldn't be defined as volatile here.

gralco committed:
Also shouldn't be defined as volatile here.

boelle commented:
the build failed, please rebase on latest devlopment

gralco commented:
Rebased.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -1770,6 +1778,26 @@ static void homeaxis(AxisEnum axis) {
+          baby_min_endstop[axis] = Y_BABY_DEFAULT_MIN_POS;
+        }
+      #endif //BABYSTEP_XY
+    #endif //BABYSTEPPING
```
Formatting should be like so:

```
    #ifdef BABYSTEPPING
      if (axis == Z_AXIS) {
        baby_max_endstop[axis] = Z_BABY_DEFAULT_MAX_POS;
        baby_min_endstop[axis] = Z_BABY_DEFAULT_MIN_POS;
      }
      #ifdef BABYSTEP_XY
        else if (axis == X_AXIS) {
          baby_max_endstop[axis] = X_BABY_DEFAULT_MAX_POS;
          baby_min_endstop[axis] = X_BABY_DEFAULT_MIN_POS;
        }
        else if (axis == Y_AXIS) {
          baby_max_endstop[axis] = Y_BABY_DEFAULT_MAX_POS;
          baby_min_endstop[axis] = Y_BABY_DEFAULT_MIN_POS;
        }
      #endif //BABYSTEP_XY
    #endif //BABYSTEPPING
```

http://www.marlinfirmware.org/index.php/Coding_Standards

Wackerbarth replied:
@thinkyhead -- We should think about using a style which is defined in terms of astyle options. Then we can use that program to enforce style in our repository.

gralco replied:
@Wackerbarth
Which program?
It would be pretty sweet if contributors could use whatever aesthetic syntax style they prefer and then it is formatted automatically upon merge/commit; does something like that exist?

Wackerbarth replied:
The program to which I was referring is "astyle". There is some danger in straying from the target style. However, this program, or one like it, can do a pretty good job. But, some manual intervention (after conversion) may be required.

gralco replied:
@Wackerbarth
Well what about simple things like curly brace placement, indentation, and whitespace? Would those ever need intervention?

Wackerbarth replied:
What else is there? (parentheses)  Don't misinterpret my comment. Generally, it does very well in converting from one style to another. However there are occasionally things that don't come out quite as expected.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -2058,6 +2086,82 @@ inline void gcode_G4() {
+        }
+      #endif //BABYSTEP_XY
+      else
+        continue;
```
This is good, but I think we can make it smaller too!

```
    const uint8_t axis_order[] = { Z_AXIS, X_AXIS, Y_AXIS };
    #ifdef BABYSTEP_XY
      for (uint8_t i=0; i<3; i++) { // in the order: Z, X, then Y
    #else
      for (uint8_t i=0; i<1; i++) { // Z only
    #endif
      boolean endstop = false;
      int8_t a = axis_order[i];
      if (code_seen[axis_codes[a]] && axis[i] == -1) {
        axis[i] = a;
        SERIAL_PROTOCOLPGM("Babystepping ");
        SERIAL_CHAR(axis_codes[a]);
        SERIAL_CHAR(' ');
      }
      else
        continue;
    . . .
```

gralco replied:
@thinkyhead
Excellent thinking!

The for loop encompasses the bulk of the _abstract_ implementation, regardless of which axis is requested. Having two separate loops like that would actually expand it.

I think this way makes sense (and was tested to work):

```
    int8_t axis[3] = { -1, -1, -1 };
    const uint8_t axis_order[] = { Z_AXIS
    #ifdef BABYSTEP_XY
      , X_AXIS, Y_AXIS }; // in the order: Z, X, then Y
    #else
      };
    #endif //BABYSTEP_XY
    for(uint8_t i=0; i<(sizeof(axis_order)/sizeof(*axis_order)); i++) {
      boolean endstop = false;
      if (code_seen(axis_codes[axis_order[i]]) && axis[i] == -1) {
        axis[i] = axis_order[i];
        SERIAL_PROTOCOLPGM("Babystepping ");
        SERIAL_CHAR(axis_codes[axis_order[i]]);
        SERIAL_CHAR(' ');
      }
      else
        continue;
. . .
```

thinkyhead commented on Marlin/ultralcd.cpp:
```diff
@@ -450,6 +450,12 @@ void lcd_set_home_offsets() {
+     #ifdef BABYSTEP_OFFSET
+      if(axis == Z_AXIS)
+        home_offset[axis] += ((int)encoderPosition)/axis_steps_per_unit[axis];
+     #endif //BABYSTEP_OFFSET
```
Call me old-fashioned, but I like doing repeated calculations up-front even though the compiler will probably optimize it into the very same code. In this case it's a little more readable too.

```
      float diff = int(encoderPosition) / axis_steps_per_unit[axis];
      baby_max_endstop[axis] -= diff;
      baby_min_endstop[axis] -= diff;
      #ifdef BABYSTEP_OFFSET
        if (axis == Z_AXIS) home_offset[Z_AXIS] += diff;
      #endif //BABYSTEP_OFFSET
```

gralco committed:
Simplified how the axis are determined.

gralco committed:
Simplified how the axis are determined.

gralco committed:
Conforming to the agreed asthetic preference. Using 'diff' per thinkyhead.

gralco committed:
Conforming to the agreed asthetic preference. Using 'diff' per thinkyhead.

gralco committed:
A little bit better this way.

gralco committed:
A little bit better this way.

gralco committed:
Discovered the need for this after additional testing.

mkile committed:
Slightly updated translation

Slightly updated translation.

mkile committed:
Slightly updated translation

Slightly updated translation.

mkile committed:
Update language_ru.h

mkile committed:
Update language_ru.h

gralco committed:
Fix additional conflicts.

thinkyhead committed:
Rename SDCARDDETECT to SDCARDDETECT_PIN

- Rename the pin so it can be tested with `PIN_EXISTS`
- Fix some incorrect tests for `SDCARDDETECT`

thinkyhead committed:
Tweaks to pins comments

thinkyhead committed:
SDCARDDETECT -> SD_DETECT, added docs

- `SD_DETECT_PIN` replaces `SDCARDDETECT`
- `SD_DETECT_INVERTED` replaces `SDCARDDETECTINVERTED`
- Revise the description of `SD_DETECT_INVERTED`
- Add a note about the override of `SD_DETECT_INVERTED` in
`Conditionals.h`

thinkyhead committed:
In for a pound, spacing adjustments

thinkyhead committed:
Add sanity checks for old names

thinkyhead committed:
Group the fail case in one negation

thinkyhead committed:
Merge pull request #2572 from thinkyhead/autostart_tweak

Tweak logic in checkautostart

Wackerbarth committed:
Development Branch DEPRECATED

thinkyhead committed:
Merge pull request #2569 from thinkyhead/rename_sdcarddetect

Rename SDCARDDETECT to SD_DETECT_PIN

thinkyhead committed:
Merge pull request #2579 from djabi/Development

Development

gralco committed:
Use if ENABLED(BABYSTEP...)

gralco committed:
Merge branch 'G5-X-Y-Z-Babystepping' of https://github.com/gralco/Marlin into HEAD
